lib/semantics/resolve-names.cc

```diff
@@ -4850,6 +4847,146 @@ bool ResolveNamesVisitor::BeginScope(const ProgramTree &node) {
   }
 }
 
+// Perform checks that need to happen after all of the specification parts
+// but before any of the execution parts.
+void ResolveNamesVisitor::FinishSpecificationParts(const ProgramTree &node) {
+  if (!node.scope()) {
+    return;  // error occurred creating scope
+  }
+  SetScope(*node.scope());
+  for (Scope &childScope : currScope().children()) {
+    if (childScope.kind() == Scope::Kind::DerivedType && childScope.symbol()) {
+      FinishDerivedType(childScope);
+    }
+  }
+  for (const auto &child : node.children()) {
+    FinishSpecificationParts(child);
+  }
+}
+
+static int FindIndexOfName(
+    const SourceName &name, std::vector<Symbol *> symbols) {
+  for (std::size_t i{0}; i < symbols.size(); ++i) {
+    if (symbols[i] && symbols[i]->name() == name) {
+      return i;
+    }
+  }
+  return -1;
+}
+
+// Perform checks on procedure bindings of this type
+void ResolveNamesVisitor::FinishDerivedType(Scope &scope) {
```

Comment:
Similarly, I think the declaration here can be ```FinishDerivedType(const Scope &scope)```.

Reply:
It would still compile if scope were declared const, but it doesn't seem right to me. I think you should only be able to get to const symbols through a const scope, and this function needs mutable symbols.

This may be a short-coming in the interface of Scope, but I don't know how to fix it.

Reply:
Sounds good.  Thanks for the explanation.